appinventor/components/src/com/google/appinventor/components/runtime/Navigation.java

```diff
@@ -0,0 +1,352 @@
+      if (method.equals(t.method())) {
+        this.method = t;
+      }
+    }
+  }
+
+  @SimpleProperty(category = PropertyCategory.BEHAVIOR,
+      description = "The longitude of the end location.")
+  public String TransportationMethod() {
+    return method.method();
+  }
+
+  @SimpleFunction(description = "Set the end location.")
+  public void SetEndLocation(double latitude, double longitude) {
+    if (!isValidLatitude(latitude)) {
+      getDispatchDelegate().dispatchErrorOccurredEvent(this, "SetEndLocation",
+          ErrorMessages.ERROR_INVALID_LATITUDE, latitude);
+    } else if (!isValidLongitude(longitude)) {
+      getDispatchDelegate().dispatchErrorOccurredEvent(this, "SetEndLocation",
+          ErrorMessages.ERROR_INVALID_LONGITUDE, longitude);
+    } else {
+      endLocation.setCoords(latitude, longitude);
+    }
+  }
+
+  private Double[][] getCoordinates() {
+    Double[][] coords = new Double[2][2];
+    coords[0][0] = startLocation.getLongitude();
+    coords[0][1] = startLocation.getLatitude();
+    coords[1][0] = endLocation.getLongitude();
+    coords[1][1] = endLocation.getLatitude();
+    return coords;
+  }
+
+  private String getCoordinatesFromGeoJson(String geoJson) {
+    String coordString = "";
+    try {
+      ArrayList<Object> o = (ArrayList<Object>) JsonUtil.getObjectFromJson(geoJson);
+      for (Object x : o) {
+        ArrayList<Object> entry = (ArrayList<Object>) x;
+        if (entry.get(0).toString().equals("features")) {
+          ArrayList<Object> feature = (ArrayList<Object>) ((ArrayList<Object>) entry.get(1)).get(0);
+          for (Object y : feature) {
+            ArrayList<Object> attr = (ArrayList<Object>) y;
+            if (attr.get(0).toString().equals("geometry")) {
+              ArrayList<Object> geometry = (ArrayList<Object>) attr.get(1);
+              for (Object z : geometry) {
+                attr = (ArrayList<Object>) z;
+                if (attr.get(0).toString().equals("coordinates")) {
+                  ArrayList<Object> coordList = (ArrayList<Object>) attr.get(1);
+                  coordString = JsonUtil.getJsonRepresentation(coordList);
+                }
+                break;
+              }
+              break;
+            }
+          }
+          break;
+        }
+      }
+    } finally {
+      return coordString;
+    }
+  }
+
+  /**
+   * Event indicating that a request has finished and has returned data (translation).
+   *
+   * @param responseCode the response code from the server
+   * @param directions the response content from the server
+   */
+  @SimpleEvent(description = "Event triggered when the Yandex.Translate service returns the " +
+      "translated text. This event also provides a response code for error handling. If the " +
+      "responseCode is not 200, then something went wrong with the call, and the translation will " +
+      "not be available.")
+  public void GotDirections(String directions) {
```

Comment:
Rather than returning the directions as a string containing a list of points, I think it would be better if we returned a YailList with (lat, long) pairs that can be passed directly into LineString's Points method, such that an event handler might be:

```
when Navigation1.GotDirections(directions)
do:
    set LineString1.Points to directions
```